File: src/sql/postgres/types/PostgresString.zig
pub const to = 25;
pub const from = [_]short{1002};

pub fn toJSWithType(
    globalThis: *jsc.JSGlobalObject,
    comptime Type: type,
    value: Type,
) AnyPostgresError!JSValue {
    switch (comptime Type) {
        [:0]u8, []u8, []const u8, [:0]const u8 => {
            var str = bun.String.borrowUTF8(value);
            defer str.deinit();
            return str.toJS(globalThis);
        },

        bun.String => {
            return value.toJS(globalThis);
        },

        *Data => {
            var str = bun.String.borrowUTF8(value.slice());
            defer str.deinit();
            defer value.deinit();
            return str.toJS(globalThis);
        },

        else => {
            @compileError("unsupported type " ++ @typeName(Type));
        },
    }
}

pub fn toJS(
    globalThis: *jsc.JSGlobalObject,
    value: anytype,
) !JSValue {
    var str = try toJSWithType(globalThis, @TypeOf(value), value);
    defer str.deinit();
    return str.toJS(globalThis);
}

const bun = @import("bun");
const AnyPostgresError = @import("../AnyPostgresError.zig").AnyPostgresError;
const Data = @import("../../shared/Data.zig").Data;

const int_types = @import("./int_types.zig");
const short = int_types.short;

const jsc = bun.jsc;
const JSValue = jsc.JSValue;

File: src/sql_jsc/postgres/types/PostgresString.rs
use crate::jsc::{JSGlobalObject, JSValue, StringJsc as _, js_error_to_postgres};
use bun_sql::postgres::AnyPostgresError;
use bun_sql::shared::Data;

/// "no impl" compile error.
pub trait ToJsWithType {
    fn to_js_with_type(self, global: &JSGlobalObject) -> Result<JSValue, AnyPostgresError>;
}

impl ToJsWithType for &[u8] {
    fn to_js_with_type(self, global: &JSGlobalObject) -> Result<JSValue, AnyPostgresError> {
        let str = bun_core::String::borrow_utf8(self);
        // `defer str.deinit()` → Drop on bun_core::String
        str.to_js(global).map_err(js_error_to_postgres)
    }
}

impl ToJsWithType for bun_core::String {
    fn to_js_with_type(self, global: &JSGlobalObject) -> Result<JSValue, AnyPostgresError> {
        self.to_js(global).map_err(js_error_to_postgres)
    }
}

// Takes owned `Data`; Drop at the
// end of this fn frees it (same pattern as bytea.rs/json.rs).
impl ToJsWithType for Data {
    fn to_js_with_type(self, global: &JSGlobalObject) -> Result<JSValue, AnyPostgresError> {
        let str = bun_core::String::borrow_utf8(self.slice());
        // `defer str.deinit()` → Drop on bun_core::String. `to_js` copies into
        // a JSC-owned string, so dropping `self` (the Data) on return is safe.
        str.to_js(global).map_err(js_error_to_postgres)
    }
}
